tesseract_environment/src/ofkt/ofkt_state_solver.cpp

```diff
@@ -291,9 +304,11 @@ void OFKTStateSolver::onEnvironmentChanged(const Commands& commands)
 
     switch (command->getType())
     {
-      case tesseract_environment::CommandType::ADD:
+      case tesseract_environment::CommandType::ADD_LINK:
       {
-        const auto& cmd = static_cast<const tesseract_environment::AddCommand&>(*command);
+        const auto& cmd = static_cast<const tesseract_environment::AddLinkCommand&>(*command);
+        assert(cmd.getJoint() != nullptr);
+
```

Comment:
What about adding a ReplaceLinkCommand?

Reply:
We could add another command. I figured it would probably be easier to just add a flag to the existing one that defaults to the current behavior.

Reply:
Actually, I just went through the code and I think your approach would be best. Adding the replace command would also have the issue with what do you do if the link does not exist. I will add a flag that enables the ability to replace link if it exist but it will use the existing joint if a joint was not provided. Sound good?

Reply:
I think that sounds good. Is this what you were thinking?
- Link provided, link already exists -> replace link
- Link provided, link doesn't exist -> add link
- Joint provided, joint already exists -> replace joint
- Joint provided, joint doesn't exist -> add joint
- Joint not provided, joint already exists -> use existing joint
- Joint not provided, joint doesn't exist -> error?

Reply:
This is what I have implemented so far. I currently writing unit tests to make sure things are working as expects. With these changes the `ADD_LINK` does not accurately represent the command not that it can be used to replace only a joint should it be change?

| Link (Exists) | Joint  (Exists) | Description |
| ---| --- | --- |
| Provided (Yes) | Not Provided | replace link use existing joint |
| Provided (No) | Not Provided | add link with joint attaching to root link |
| Not Provided | Provided (Yes) | replace joint which may move if the parents are not the same |
| Not Provided | Provided (No) | error |
| Provided (Yes) | Provided (Yes) | replace link and joint (These do not have to be associated in any way) |
| Provided (No) | Provided (No) | add link and joint (joint child must be the provided link) |

Reply:
We could move the logic for the 3rd and 4th row into the move joint command. That might fit the description a bit better

Reply:
Well during the process it was getting messy so I broke out the replace joint into its own command

AddLinkCommand:

| Link (Exists) | Joint  (Exists) | Description |
| ---| --- | --- |
| Provided (Yes) | Not Provided | replace link use existing joint |
| Provided (No) | Not Provided | add link with joint attaching to root link |
| Provided (No) | Provided (No) | add link and joint (if child link is not the name of the provided link it errors) |
| Provided (Yes) | Provided (No) | error |
| Provided (No) | Provided (Yes) | error |

ReplaceJointCommand:

| Joint  (Exists) | Description |
| ---| --- |
| Provided (Yes) | replace joint use existing joint (if child link is not the same it results in an error) |
| Provided (No) | error |

Reply:
@mpowelson I just created a [PR](https://github.com/ros-industrial-consortium/tesseract/pull/547) add support for replacing links and joints. It was a little more work than I expected but I believe it appears to be working.